java-support/src/main/java/io/cloudstate/javasupport/CloudState.java

```diff
@@ -300,6 +304,75 @@ public CloudState registerAction(
     return this;
   }
 
+  /**
+   * Register a annotated value entity.
```

Comment:
```suggestion
   * Register an annotated value entity.
```

Reply:
@pvlugter if we want to use `Entity` in the language support and in the proxy.  I think, the doc should be `Register an annotated entity.`. If I understand correctly it is just about renaming `ValueEntity` to `Entity` everywhre. I just want to be sure.

Reply:
@ralphlaude @pvlugter I'm curious if ValueEntity (previous CRUD) is a ValueEntity or now an Entity, then how does this entity discriminate itself from an "eventsourced" **entity**, a "crdt" **entity** and an "action"(?) **entity** by its naming?

I'm writing this as I found having packages for each state model in Go, where those entities are categorized into being: `eventsourced`, `crdt` and `action`. Similar as we have `valueentity/ValueEntity.java` I would have to have an `entity.Entity`as a type, which seems to be a generalized categorization by its name alone. How does that fit into the different state models we have? How is an "entity Entity" a state model?

Reply:
I wonder if we made a good choice of names. Initially this was supposed to be a CrudEntity that would be an Entity capable of doing simple CRUD operations that would be implemented on top of EventSourced support (which seems to me to be an EventSourced specialization), however it became a generic and default entity called Entity or ValueEntity which seemed very confusing. I think we should be explicit about the names of the entity and not support a generic or default entity because the user should want to know what he is doing and nothing better than a name that clearly indicates this.

Reply:
I am inclined to agree with @sleipnir and @marcellanz here. `ValueEntity` is ambiguous for several reasons (because the word Value is) and because it doesn't really say anything about the state model.

Alternatives:

  * `ReadWriteEntity`
  * `SnapshotEntity`
  * `MutableEntity`
  * `FriendlyNeighbourhoodEntity`
  * and full-circle: `CRUDEntity`

Reply:
If I read `protocols/protocol/cloudstate/value_entity.proto` right, we're pushing around state, created initially by the user function and then let it persisted by the proxy. Beside mutating that state, and deleting it, the proxy does not know anything about that state. What is that called?

Reply:
> Alternatives:
> ...
> * `FriendlyNeighbourhoodEntity`
> * and full-circle: `CRUDEntity`

@viktorklang I like the extreme value analysis approach here.

Reply:
`CRUD Entity` didn't seem like the best name, given that there shouldn't be explicit `create` or `read` operations — entities are implicitly created when accessed and with the Cloudstate inversion, the state is passed into the entity, not fetched. And while it's a technical acronym, it's worth noting that the word `crud` also has negative connotations (dictionary definition: _a substance which is considered unpleasant or disgusting, typically because of its dirtiness_).

Discussions around `Key-Value` led to this being just a `Value` based state model for entities, since the key is always part of the entity itself and not explicitly in its state model. But we can keep looking for a better name for this simple state model.

In terms of making it the default, when @jroper and I discussed, the thinking was that the value-based state model was the simplest and basically synonymous with the core functionality of entities. Where an entity has a distinct identity, associated with its id/key, and has a state. Just updating its state directly is the simplest model, and this could be the expected default for an `Entity`. While the event-sourced and replicated entities are specialised entities. So that there's `Entity` (with a simple set-value or set-state function), `Event-Sourced Entity` (tracks state using event sourcing, instead of just latest snapshot), and `Replicated Entity` (replicates the state across instances). And that Actions are not Entities, they don't have a distinct identity, they're just Actions.

Reply:
@pvlugter Isn't `getState` a read method and `updateState` is the create method (if the state was initially Optional.empty)?

I think we need to approach the naming from what we believe users will have an intuition for, knowing that ValueEntity has some form of etymological story coming from Key-Value Entity seems like a bit of a stretch (and to my knowledge has no "prior art").

Reply:
That said "Entity" might be fine, alas it renders it difficult to talk about "different kinds of Entities".

`DefaultEntity` might be more suitable if it indeed is intended to be the default state model.

Reply:
To be aligned here with other existing state models and also to be explicit to the user the only name left is `CrudEntity`. We already have `EventSourcedEntity` and `CrdtEntity`. They convey the meaning and what the user should expect and they are not generic. `Entity`, `ValueEntity` and `DefaultEntity` are generic and do not express the intent of the state model.

This entity is implemented by passing the state around and mutating the state. The proxy initially fetches the state and pass it around. Event Sourcing is not involved here anymore.

Reply:
> Isn't `getState` a read method and `updateState` is the create method (if the state was initially Optional.empty)?

I think conceptually we were thinking more along the lines of the original Cloudstate ideas, where rather than calling a state management API, the state is passed to the function — so it's a function that receives the request and state, and returns the response and updated state, and the focus is on state transformation. There may be a `get` method, or the state gets passed in. There may be an `update` method, or the updated state is returned. The main operation for the user conceptually would be _update_. The _create_, _read_, _update_, _delete_ functions are of course all there for persistent storage in general, but even in the simple model there's also the inversion, and where parts of this are managed automatically.

Reply:
When discussing making this the default, we were expecting that the user APIs would just use `Entity`. So that it's something like this with annotations:

```
@Entity
class MyEntityType
```
_with (insert name) default simple state management: state transformation command functions or `update` method_

```
@EventSourcedEntity
class MyEntityType
```
_with snapshot and event handlers and event emitting APIs_

```
@ReplicatedEntity
class MyEntityType
```
_with the CRDT data type APIs_

Reply:
Would be good to have a clear name for the simple state management model. Personally, I'm not sold on `CRUD` for the name. I do think just `Entity` works in the user APIs. While there might not be precedent for `Value` based, I think it fitted quite well in the initial discussions, but certainly no need to keep it. There are different interpretations for _value_ of course, but I think the general computer science usage is okay here:

> In computer science, a value is the representation of some entity that can be manipulated by a program.

https://en.wikipedia.org/wiki/Value_(computer_science)

Reply:
I discussed this topic with @marcellanz  and we came to the conclusion that all the names created until today implied expected implicit behavior. If the purpose of this entity is, as referenced here sometimes, to be just "an object subject to mutation that will have its state stored" I think that **MutableEntity** seems to be the most appropriate. This entity is very similar to the concept of _Document-oriented databases of documents_ and we came, Marcel and I, to think of calling this **DocumentEntity** but that would not manifest the behavior of itself and therefore we think it better to find some name that denotes the behavior in a more specific way and I think **MutableEntity** (given what everything they said about what this entity seems to do) seems to be a suitable name in my opinion. This seems to indicate that given an initial state, mutation operations of that state can subsequently be applied to that entity.

Reply:
I don't know if Document is the right term, alas I brought it to describe the "data as a whole" nature of such an entity. Also Document has its associations with some databases I think.

I see that its an entity with nothing more than representing a value that can be set, mutaded and deleted. It could be immuntable too, if that helps the proxy to do anything useful with that hint of its mutability.

Reply:
>https://en.wikipedia.org/wiki/Value_(computer_science)

I think this is a case of "technically right but practically wrong"—I think what matters is what will seem "intuitive" to users/developers. Just as CRDT is technically correct but will be renamed to something like Replicated since it is more likely tractable with the target audience.

Getting the current state as a parameter (Optional) and return a response + a new state (Optional) will be a bit of a departure from the current APIs—as they tend to use the Context parameter to manipulate their state (emit etc).

While I am not against doing it that way, I do believe that users will be more inclined to use mutation rather than "pure" behaviors (for the majority of the languages we intend to support).

In order to make progress, I believe we should take a bit of guidance decisions:

  1) What should be conveyed by the name of the type of entity?
  2) What do we need to consider in terms of API-possibility which would make it possible to convert from one state representation to another: CRUD -> EventSourcing, CRUD -> CRDT, etc
  3) Should User-facing APIs be idiomatic to the language/setting in which they are used? (I believe we agree that the answer here is yes, but want to make sure)
 

As for `MutableEntity` I think that's an OK name, but it sort of implies that the other state models aren't—which is not the case. Perhaps `UpdateableEntity` would be more appropriate to distinguish?

Reply:
@viktorklang In fact, EventSourced technically always presents a new state, apparently "mutable", but it is always a new non-destructible state. And CRDT is not exactly about state mutability, but how to converge to a quorum state in a distributed environment. 

I agree on defining the guidance issues.

Reply:
How about `PersistentEntity`?

This is still more of a default term (other entity types are persistent too) but it focuses on the main distinction of what it does over a plain, vanilla `Entity` (which could have non-persistent values).

And _value_ can stay as the general concept for an entity's value, which applies to all the entity types.

(Note: `PersistentEntity` was also used in [Lagom](https://www.lagomframework.com/documentation/1.6.x/scala/PersistentEntity.html) for what is an `EventSourcedEntity` in Cloudstate. I don't see any issue with using it differently though.)

Reply:
@pvlugter Wouldn't that mean that the other's are not persistent? Crdts are not (yet), but should then not EventSourcedEntity be EventSourcedPersistentEntity?

Reply:
DurableEntity is another option together with DurableEventSourcedEntity, and eventually DurableReplicatedEntity?

Reply:
Clearly, a user should not have to think or read more about this (simple?) state model type than she has to for replicated or eventsourced entities. As an example, I'd have not to explain anyone what a KV entity is.

ManagedEntity, all my JPA friends will be happy :)
DataEntity, everything is data, right?
PeristedEntity, we never forget
DurableEntity, we never forget (2)
StateEntity, whatever you think it is.

Guidance:
1: I think it should describe what the proxy does in term of its "inversion of state" duties.
2: Don't know but I found it kind of interesting how little the proxy knows about an entities state structure being just proto.Any blobs (except for certain CRDTs for merging).
3: Yes. One example of this is 'ctx.Fail' is missing in the (new) Go API.

Reply:
@pvlugter @viktorklang you really didn't came to the durable and persisted entity while I wrote that comment at the same time (into my little iPhone Github App), no? 😅

Reply:
@marcellanz Great minds think alike. Cloudstate community ftw!

Reply:
@marcellanz 

>Guidance:
1: I think it should describe what the proxy does in term of its "inversion of state" duties.

This is my inclination as well.

>2: Don't know but I found it kind of interesting how little the proxy knows about an entities state structure being just proto.Any blobs (except for certain CRDTs for merging).

I find it rather amazing really—how much value it can provide without having to know about the *content*

>3: Yes. One example of this is 'ctx.Fail' is missing in the (new) Go API.

Good point